controller/api/destination/endpoint_profile_translator.go

```diff
@@ -1,69 +1,136 @@
+
+	updates chan *watcher.Address
+	stop    chan struct{}
+
+	current *pb.DestinationProfile
 
 	log *logging.Entry
 }
 
-// newEndpointProfileTranslator translates pod updates and protocol updates to
+// endpointProfileUpdatesQueueOverflowCounter is a prometheus counter that is incremented
+// whenever the profile updates queue overflows.
+//
+// We omit ip and port labels because they are high cardinality.
+var endpointProfileUpdatesQueueOverflowCounter = promauto.NewCounter(
+	prometheus.CounterOpts{
+		Name: "endpoint_profile_updates_queue_overflow",
+		Help: "A counter incremented whenever the endpoint profile updates queue overflows",
+	},
+)
+
+// newEndpointProfileTranslator translates pod updates and profile updates to
 // DestinationProfiles for endpoints
 func newEndpointProfileTranslator(
 	enableH2Upgrade bool,
 	controllerNS,
 	identityTrustDomain string,
 	defaultOpaquePorts map[uint32]struct{},
-	log *logging.Entry,
 	stream pb.Destination_GetProfileServer,
+	endStream chan struct{},
+	log *logging.Entry,
 ) *endpointProfileTranslator {
 	return &endpointProfileTranslator{
 		enableH2Upgrade:     enableH2Upgrade,
 		controllerNS:        controllerNS,
 		identityTrustDomain: identityTrustDomain,
 		defaultOpaquePorts:  defaultOpaquePorts,
-		stream:              stream,
-		log:                 log.WithField("component", "endpoint-profile-translator"),
+
+		stream:    stream,
+		endStream: endStream,
+		updates:   make(chan *watcher.Address, updateQueueCapacity),
+		stop:      make(chan struct{}),
+
+		log: log.WithField("component", "endpoint-profile-translator"),
 	}
 }
 
+// Start initiates a goroutine which processes update events off of the
+// endpointProfileTranslator's internal queue and sends to the grpc stream as
+// appropriate. The goroutine calls non-thread-safe Send, therefore Start must
+// not be called more than once.
+func (ept *endpointProfileTranslator) Start() {
+	go func() {
+		for {
+			select {
+			case update := <-ept.updates:
+				ept.update(update)
+			case <-ept.stop:
+				return
+			}
+		}
+	}()
+}
+
+// Stop terminates the goroutine started by Start.
+func (ept *endpointProfileTranslator) Stop() {
+	close(ept.stop)
+}
+
 // Update sends a DestinationProfile message into the stream, if the same
 // message hasn't been sent already. If it has, false is returned.
 func (ept *endpointProfileTranslator) Update(address *watcher.Address) (bool, error) {
+	select {
+	case ept.updates <- address:
+		// Update has been successfully enqueued.
+	default:
+		// We are unable to enqueue because the channel does not have capacity.
+		// The stream has fallen too far behind and should be closed.
+		endpointProfileUpdatesQueueOverflowCounter.Inc()
+		select {
+		case <-ept.endStream:
+			// The endStream channel has already been closed so no action is
+			// necessary.
+		default:
+			ept.log.Error("Profile update queue full; aborting stream")
+			close(ept.endStream)
+		}
+	}
+	return true, nil
```

Comment:
The pod watcher calls this function whenever the pod changes. The boolean returned is supposed to tell whether that change implied sending an update back into the stream, in which case the pod watcher will increment the `ip_port_updates` metric. Always returning true here will inflate that accounting.

Reply:
I've updated the Update interface so now that we only return an error -- with the buffer we can't actually know whether the consumer sends the update or not. We now return an error when the address can't be enqueued.